core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -184,11 +184,23 @@ private[spark] class TaskSetManager(
     t.epoch = epoch
   }
 
+  // An array to store preferred location and its task index
+  private val locationWithTaskIndex: ArrayBuffer[(String, Int)] = new ArrayBuffer[(String, Int)]()
+  private val addTaskStartTime = System.nanoTime()
   // Add all our tasks to the pending lists. We do this in reverse order
   // of task index so that tasks with low indices get launched first.
   for (i <- (0 until numTasks).reverse) {
-    addPendingTask(i)
+    addPendingTask(i, true)
   }
+  // Convert preferred location list to rack list in one invocation and zip with the origin index
+  private val rackWithTaskIndex = sched.getRacksForHosts(locationWithTaskIndex.map(_._1).toList)
```

Comment:
would it be worth doing some de-duping here?  It would not be unusual to have a taskset with 10K tasks on 100 hosts.  I see that `CachedDNSToSwitchMapping` will cache the lookup itself, but doesn't seem necessary to create those intermediate data structures with 10K elements.

Reply:
btw I meant something like this for avoiding leaving around `locationWithTaskIndex` and doing the deduping. 
 https://github.com/squito/spark/commit/fc0b3089efe4bf35f145e5c01bff26e7b9c5f0e7

 Feel free to pull in that commit to your change if you think its helpful.

The de-duping thing is minor, but I am concerned that the `locationWithTaskIndex` variable is going to be confusing if its left around as a private member variable, even though its only meaningful in this limited context.

Reply:
> would it be worth doing some de-duping here? It would not be unusual to have a taskset with 10K tasks on 100 hosts. I see that `CachedDNSToSwitchMapping` will cache the lookup itself, but doesn't seem necessary to create those intermediate data structures with 10K elements.

We can **not** de-duping this List, or the logic about choosing a task to launch will incorrect.

Reply:
> The de-duping thing is minor, but I am concerned that the `locationWithTaskIndex` variable is going to be confusing if its left around as a private member variable, even though its only meaningful in this limited context.

Yes, I don't want to do any de-duping here. I will refactor this part for getting more readable.

Reply:
Refactored

Reply:
On de-duping -- I know you can't do it blindly, as then you won't correctly keep track of which task goes with which host.  But if you look at my proposed change, I think it keeps track of the correct mapping back to the original task indices (it certainly should be possible, anyway, even if that exact version doesn't work).

Looking at `CachedDNSToSwitchMapping`, if you repeat the same host 1000 times, if that host is not in the cache yet, it will repeat that lookup 1000 times.  [`getUncachedHosts`](https://github.com/apache/hadoop/blob/ebb5fa115b842a3bcb747e881ca2a8fe4c716d07/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/net/CachedDNSToSwitchMapping.java#L116) does a lookup to check if its in the cache, but if not it'll just pass each copy on to the `uncachedHosts` variable.  Then that just gets passed through to [`rawMapping.resolve()`](https://github.com/apache/hadoop/blob/ebb5fa115b842a3bcb747e881ca2a8fe4c716d07/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/net/CachedDNSToSwitchMapping.java#L119).  (Perhaps `getUncachedHosts` really should return a `Set<String>`.)

so after looking at that code, I'm more convinced de-duping would help you.

Reply:
> if you repeat the same host 1000 times, if that host is not in the cache yet, it will repeat that lookup 1000 times.

Looking up a host in a Set has no different with looking up it in a HashMap. For example, if host1, host2, host3 are three uncached unresolved hosts. First time, `getUncachedHosts` will return a list includes all of them. Then `rawMapping.resolve()` resolve them and `cacheResolvdHosts(uncachedHosts, resolvedHosts)` will cache the three hosts in memory. Now the second round with host1, host2, host3, host4 begin, `getUncachedHosts` will look up in a HashMap and return a list only contains host4, then the host4 will be resolved and cache in memory.
```java
  @Override
  public List<String> resolve(List<String> names) {
    // normalize all input names to be in the form of IP addresses
    names = NetUtils.normalizeHostNames(names);

    List <String> result = new ArrayList<String>(names.size());
    if (names.isEmpty()) {
      return result;
    }
    List<String> uncachedHosts = getUncachedHosts(names);
    // Resolve the uncached hosts
    List<String> resolvedHosts = rawMapping.resolve(uncachedHosts);
    //cache them
    cacheResolvedHosts(uncachedHosts, resolvedHosts);
    //now look up the entire list in the cache
    return getCachedHosts(names);
  }

  private List<String> getCachedHosts(List<String> names) {
    List<String> result = new ArrayList<String>(names.size());
    // Construct the result
    for (String name : names) {
      String networkLocation = cache.get(name);
      if (networkLocation != null) {
        result.add(networkLocation);
      } else {
        return null;
      }
    }
    return result;
  }
```

Reply:
Refactor with the de-duping approach.